Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1

```diff
@@ -473,6 +479,83 @@ function Get-MsiAccessToken {
+        $timeDifference = $tokenExpiryTime - $currentTime
+
+        if ($timeDifference.Minutes -gt 5) {
+            Write-Verbose "Returning access token from cache."
+            return $script:certificateAccessToken
+        }
+    }
+
+    Write-Verbose "Fetching access token using client certificate."
+    
+    # load the ADAL library 
+    Add-Type -Path $PSScriptRoot\Microsoft.IdentityModel.Clients.ActiveDirectory.dll
+
+    $pemFileContent = $endpoint.Auth.Parameters.ServicePrincipalCertificate
+    $pfxFilePath, $pfxFilePassword = ConvertTo-Pfx -pemFileContent $pemFileContent
+    
+    $clientCertificate = Get-PfxCertificate -pfxFilePath $pfxFilePath -pfxFilePassword $pfxFilePassword
+
+    $servicePrincipalId = $endpoint.Auth.Parameters.ServicePrincipalId
+    $tenantId = $endpoint.Auth.Parameters.TenantId
+    $envAuthUrl = $script:defaultEnvironmentAuthUri
+    if($endpoint.Data.environmentAuthorityUrl) {
+        $envAuthUrl = $endpoint.Data.environmentAuthorityUrl
+    }
+
+    $envAuthUrl = Get-EnvironmentAuthUrl -endpoint $endpoint
+    $azureActiveDirectoryResourceId = Get-AzureActiverDirectoryResourceId -endpoint $endpoint
+
+    # add check -- if envAuthUrl ends with adfs then ignore tenantId
+    $authorityUrl = "$envAuthUrl$tenantId"
+    
+    try {
+        $clientAssertionCertificate = New-Object Microsoft.IdentityModel.Clients.ActiveDirectory.ClientAssertionCertificate -ArgumentList $servicePrincipalId, $clientCertificate
+        $authenticationContext = New-Object Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContext -ArgumentList $authorityUrl
+        $tokenResult = $authenticationContext.AcquireTokenAsync($azureActiveDirectoryResourceId, $clientAssertionCertificate).ConfigureAwait($false).GetAwaiter().GetResult()
+    }
+    catch {
+        $script:certificateAccessToken = $null
+        throw (Get-VstsLocString -Key "AZ_SPNCertificateAccessTokenFetchFailure" -ArgumentList $servicePrincipalId, $_)
+    }
+
+    if ($tokenResult) {
+        Write-Verbose "Successfully fetched access token using client certificate."
+        $script:certificateAccessToken = @{
+            token_type = $tokenResult.AccessTokenType;
+            access_token = $tokenResult.AccessToken;
+            expires_on = $tokenResult.ExpiresOn;
+        }
+
+        return $script:certificateAccessToken
+    }
+    else {
+        $script:certificateAccessToken = $null
+        throw (Get-VstsLocString -Key "AZ_SPNCertificateAccessTokenFetchFailureTokenNull" -ArgumentList $servicePrincipalId)
+    }
+}
+
+function Get-PfxCertificate {
+    param(
+        [string][Parameter(Mandatory=$true)] $pfxFilePath, 
+        [string][Parameter(Mandatory=$true)] $pfxFilePassword
+    )
+
+    $clientCertificate = New-Object System.Security.Cryptography.X509Certificates.X509Certificate2
+    $clientCertificate.Import($pfxFilePath, $pfxFilePassword, [System.Security.Cryptography.X509Certificates.X509KeyStorageFlags]::PersistKeySet)
```

Comment:
When is this removed?

Reply:
Here, we only create a certificate object and not store it in the certificate store. So I believe that there is nothing to be removed here.